src/dmd/dmodule.d

```diff
@@ -653,188 +653,233 @@ extern (C++) final class Module : Package
-             * 00 00 FE FF  UTF-32BE, big-endian
-             * FF FE 00 00  UTF-32LE, little-endian
-             * FE FF        UTF-16BE, big-endian
-             * FF FE        UTF-16LE, little-endian
-             * EF BB BF     UTF-8
-             */
-            uint le;
-            uint bom = 1; // assume there's a BOM
-            if (buf[0] == 0xFF && buf[1] == 0xFE)
+            static if (endian == Endian.little) alias readNext = Port.readlongLE;
+            else alias readNext = Port.readlongBE;
+
+            if (buf.length & 3)
+            {
+                error("odd length of UTF-32 char source %u", buf.length);
+                fatal();
+            }
+
+            const (uint)[] eBuf = cast(const(uint)[])buf;
+
+            OutBuffer dbuf;
+            dbuf.reserve(eBuf.length);
+
+            foreach (i; 0 .. eBuf.length)
             {
-                if (buflen >= 4 && buf[2] == 0 && buf[3] == 0)
+                const u = readNext(&eBuf[i]);
+                if (u & ~0x7F)
                 {
-                    // UTF-32LE
-                    le = 1;
-                Lutf32:
-                    OutBuffer dbuf;
-                    uint* pu = cast(uint*)buf;
-                    uint* pumax = &pu[buflen / 4];
-                    if (buflen & 3)
+                    if (u > 0x10FFFF)
                     {
-                        error("odd length of UTF-32 char source %u", buflen);
+                        error("UTF-32 value %08x greater than 0x10FFFF", u);
                         fatal();
                     }
-                    dbuf.reserve(buflen / 4);
-                    for (pu += bom; pu < pumax; pu++)
+                    dbuf.writeUTF8(u);
+                }
+                else
+                    dbuf.writeByte(u);
+            }
+            dbuf.writeByte(0); //add null terminator
+            return dbuf.extractSlice();
+        }
+
+        /*
+         * Convert a buffer from UTF16 to UTF8
+         * Params:
+         *    Endian = is the buffer big/little endian
+         *    buf = buffer of UTF16 data
+         * Returns:
+         *    input buffer reencoded as UTF8
+         */
+
+        char[] UTF16ToUTF8(Endian endian)(const(char)[] buf)
+        {
+            static if (endian == Endian.little) alias readNext = Port.readwordLE;
+            else alias readNext = Port.readwordBE;
+
+            if (buf.length & 1)
+            {
+                error("odd length of UTF-16 char source %u", buf.length);
+                fatal();
+            }
+
+            const (ushort)[] eBuf = cast(const(ushort)[])buf;
+
+            OutBuffer dbuf;
+            dbuf.reserve(eBuf.length);
+
+            //i will be incremented in the loop for high codepoints
+            foreach (ref i; 0 .. eBuf.length)
+            {
+                uint u = readNext(&eBuf[i]);
+                if (u & ~0x7F)
+                {
+                    if (0xD800 <= u && u <= 0xDBFF)
```

Comment:
I suggested `0xD800 <= u && u < 0xDC00`. This might seem arbitrary, but it's consistent with how loops work: `foreach (u; 0xD800 .. 0xDC00)`

Reply:
Make sense